src/buffer.c

```diff
@@ -232,6 +347,38 @@ buffer_peek(const struct Buffer *src, void *dst, size_t len) {
 
         bytes_copied += iov[i].iov_len;
     }
+#if 0
+    if (src->type == SOCK_DGRAM) {
+        if (src->len == 0)
```

Comment:
I think this should be:
```
uint16_t *dgram_len = (uint16_t *)&buffer->buffer[buffer->head];
if (src->len < 2
        || len < *dgram_len) {
    return 0;
} ...
```
I would just store the original `src->head` and `src->len` in local variables are restore them at the end.